crates/amaru-consensus/src/consensus/tx_submission/tx_submission_server_state.rs

```diff
@@ -0,0 +1,189 @@
+
+/// State of a transaction submission server for a given peer.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TxSubmissionServerState {
+    /// Server parameters: batch sizes, window sizes, etc.
+    params: ServerParams,
+    /// Peer we are connecting to.
+    peer: Peer,
+    /// All tx_ids advertised but not yet acked (and their size).
+    window: VecDeque<(TxId, u32)>,
+    /// Tx ids we want to fetch but haven't yet requested.
+    pending_fetch: VecDeque<TxId>,
+    /// Tx ids we requested; waiting for replies.
+    /// First as a FIFO queue because when we receive tx bodies we don't get the ids back.
+    inflight_fetch_queue: VecDeque<TxId>,
+    /// Then as a set for quick lookup when processing received ids.
+    /// This is kept in sync with `inflight_fetch_queue`. When we receive a tx body,
+    /// we pop it from the front of the queue and remove it from the set.
+    inflight_fetch_set: BTreeSet<TxId>,
+    /// Tx ids we processed but didn't insert (invalid, policy failure, etc.).
+    rejected: BTreeSet<TxId>,
+}
+
+impl TxSubmissionServerState {
+    pub fn new(peer: &Peer, params: ServerParams) -> Self {
+        Self {
+            params,
+            peer: peer.clone(),
+            window: VecDeque::new(),
+            pending_fetch: VecDeque::new(),
+            inflight_fetch_queue: VecDeque::new(),
+            inflight_fetch_set: BTreeSet::new(),
+            rejected: BTreeSet::new(),
+        }
+    }
+
+    pub fn peer(&self) -> &Peer {
+        &self.peer
+    }
+
+    pub fn process_tx_ids_reply(
+        &mut self,
+        mempool: Arc<dyn TxSubmissionMempool<Tx>>,
+        tx_ids: Vec<(TxId, u32)>,
+    ) -> anyhow::Result<Option<Vec<TxId>>> {
+        self.received_tx_ids(mempool, tx_ids)?;
+        Ok(self.txs_to_request())
+    }
+
+    pub async fn process_txs_reply(
+        &mut self,
+        mempool: Arc<dyn TxSubmissionMempool<Tx>>,
+        txs: Vec<Tx>,
+    ) -> anyhow::Result<(u16, u16, Blocking)> {
+        self.received_txs(mempool.clone(), txs).await?;
+        self.request_tx_ids(mempool.clone()).await
+    }
+
+    pub async fn request_tx_ids(
+        &mut self,
+        mempool: Arc<dyn TxSubmissionMempool<Tx>>,
+    ) -> anyhow::Result<(u16, u16, Blocking)> {
+        // Acknowledge everything we’ve already processed.
+        let mut ack = 0_u16;
+
+        while let Some((tx_id, _size)) = self.window.front() {
+            let already_in_mempool = mempool.contains(tx_id);
+            let already_rejected = self.rejected.contains(tx_id);
+
+            if already_in_mempool || already_rejected {
+                // pop from window and ack it
+                if let Some((front_id, _)) = self.window.pop_front() {
+                    // keep rejected set from growing forever
+                    if already_rejected {
+                        self.rejected.remove(&front_id);
+                    }
+                    ack = ack.saturating_add(1);
```

Comment:
shouldn’t we stop the loop upon saturating the u16? otherwise the internal state and the network state diverge